### PR TITLE
add initial CAA record protobuf

### PR DESCRIPTION
**Pull request by kumarde**
Feedback appreciated.

### CODE REVIEW

File: proto/caa.proto [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+message CAATagValue {
+    bool flag = 1;
+    string tag = 2;
+    string value = 3;
+    int64 ttl = 4;
+}
+
+message CAARecord {
+    int64 timestamp = 1;
```
Comment (zakird) [inline proto/caa.proto]: Does this need its own timestamp? Why doesn't the one in CAALookup suffice?

File: proto/caa.proto [hunk 2]
```diff
@@ -0,0 +1,39 @@
+syntax="proto3";
+package zsearch;
+
+enum CAAStatus {
+    CAA_STATUS_RESERVED = 0;
+    CAA_STATUS_VALIDATION_SUCCESS = 1;
+    CAA_STATUS_VALIDATION_FAILED = 2;
```
Comment (zakird) [inline proto/caa.proto]: Given that the `DNS_ERROR` names are `FAIL` and not `FAILED`, I think we might want this to be `CAA_STATUS_VALIDATION_FAIL`.

Comment (zakird) [reply proto/caa.proto]: We probably need a `SKIPPED` as well. Some lookups earlier on in processing can lead to others being skipped I _think_. Someone should check the RFC.

File: proto/caa.proto [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    CAA_STATUS_DNS_ERROR_ILLEGAL_INPUT = 9;
+    CAA_STATUS_DNS_ERROR_TIMEOUT = 10;
+    CAA_STATUS_DNS_ERROR_ITER_TIMEOUT = 11;
+    CAA_STATUS_DNS_ERROR_TEMPORARY = 12;
+    CAA_STATUS_DNS_ERROR_TRUNCATED = 13;
+}
+
+message CAATagValue {
+    bool flag = 1;
+    string tag = 2;
```
Comment (zakird) [inline proto/caa.proto]: Does this need to be a string, or can it be an ENUM? What is it in the RFC?

Comment (zakird) [reply proto/caa.proto]: (strings are pricy to store)

Comment (kumarde) [reply proto/caa.proto]: unfortunately, the RFC allows arbitrary tags

Comment (zakird) [reply proto/caa.proto]: _sigh_ ok

File: proto/caa.proto (hunk 2 above)
Comment (kumarde) [reply proto/caa.proto]: Yeah, I can add that. There are three exception cases where CAA doesn't have to be checked.

Comment (kumarde) [reply proto/caa.proto]: (That being said, not totally sure we can test that from our vantage point.)

File: proto/caa.proto (hunk 1 above)
Comment (kumarde) [reply proto/caa.proto]: I figure they happen at _slightly_ different times. But I can just keep the outer one and nix these.

**Comment by zakird** [comment]
That sounds good to me..

**Comment by zakird** [comment]
@kumarde can you merge master and rebuild?

**Comment by kumarde** [comment]
done

**Comment by zakird** [comment]
This looks good to me. Once @dadrian takes a look, we should be fine to merge.

File: .gitignore [hunk 4]
```diff
@@ -37,3 +37,7 @@
 python/build/*
 python/dist/*
 python/zsearch_definitions.egg-info/*
+
+#IntelliJ
```
Comment (dadrian) [inline .gitignore]: Why do you never put a space between comment characters and the start of a comment? 😞

File: proto/caa.proto [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    CAA_STATUS_DNS_ERROR_ITER_TIMEOUT = 12;
+    CAA_STATUS_DNS_ERROR_TEMPORARY = 13;
+    CAA_STATUS_DNS_ERROR_TRUNCATED = 14;
+}
+
+message CAATagValue {
+    bool flag = 1;
+    string tag = 2;
+    string value = 3;
+    int64 ttl = 4;
```
Comment (dadrian) [inline proto/caa.proto]: Is this an int or a uint?

File: proto/caa.proto [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+message CAARecord {
+    bytes raw = 1;
+    string domain = 2;
+    CAAStatus result = 3;
+    repeated CAATagValue values = 4;
+}
+
+message CAALookup {
+    int64 timestamp = 1;
```
Comment (dadrian) [inline proto/caa.proto]: This should probably also be unsigned?

File: proto/certificate.proto [hunk 7]
```diff
@@ -173,6 +174,7 @@ message Certificate {
     CertificateValidation validation = 35;
     CTStatus ct = 30;
     ZLint zlint = 38;
+    CAALookup caa = 49;
```
Comment (dadrian) [inline proto/certificate.proto]: Can you add a comment at the bottom of `Certificate` that has the highest ID used?

File: .gitignore (hunk 4 above)
Comment (kumarde) [reply .gitignore]: LOL I have never once thought about this...I do not have a good answer

File: proto/caa.proto (hunk 5 above)
Comment (kumarde) [reply proto/caa.proto]: Yeah, should be a uint.

File: proto/caa.proto (hunk 6 above)
Comment (kumarde) [reply proto/caa.proto]: fixed

File: proto/certificate.proto (hunk 7 above)
Comment (kumarde) [reply proto/certificate.proto]: done

**Pull request merged**